aristanetworks/ansible-cvp
# Feat(module_utils): Add support for approve/execute actions

## Change Summary

Address the feature request to add support for approve/execute of a CVP change controls in the cv_change_control_v3 module.

## Related Issue(s)

Fixes #526 

## Component(s) name

`arista.cvp.cv_change_control_v3

## Proposed changes
Add support for following actions in cv_change_control_v3:
- "approve": approve a single change control in CVP
- "unapprove": unapprove a single change control in CVP
- "execute": start execution of a single change control in CVP
- "approve_and_execute": approves and starts execution of a single change control in CVP

The new actions make use of cvprac resource APIs and largely follow the pattern of the existing actions. Note that included in the proposed changes are updates to replace some existing calls to the post wrapper with calls to cvprac resource APIs.

Other implementation notes:
- approve/execute actions only support handling one CC at a time, attempts to initiate action on multiple CCs will be rejected
- added cv_change_control_v3 to the list of scope for the PR triage workflow

## How to test
The example playbooks form the basis for how to test these changes. Note that examples for create/show/remove were provided to help verify the changes to existing code to leverage cvprac resource APIs.

For approve/execute actions the most basic test is to create a change control and then apply approve/execute actions. If CC execution was started as expected then the final show of the CC will demonstrate that it is in running or completed state.

```
---
- name: CVP Change Control Test approve_and_execute
  hosts: cv_server
  gather_facts: no
  vars:
    change:
      name: Leaf 1A Change Control
      notes: Created via playbook
      activities:
        - action: "mlaghealthcheck"
          name: Check_LEAF1A_MLAG_Health
          arguments:
            - name: DeviceID
              value: SN-DC1-POD1-LEAF1A
          stage: LEAF1A_MLAG_Health
      stages:
        - name: LEAF1A_MLAG_Health
          mode: parallel

  tasks:
    - name: "Create a change control on {{inventory_hostname}}"
      arista.cvp.cv_change_control_v3:
        state: set
        change: "{{ change }}"
      register: cv_change_control_test

    - name: "Approve and Execute a change control on {{inventory_hostname}} using change_id field"
      arista.cvp.cv_change_control_v3:
        state: approve_and_execute
        change_id: ["{{ cv_change_control_test.data.id }}"]

    - name: "Show the change control on {{inventory_hostname}} using CC name"
      arista.cvp.cv_change_control_v3:
        state: show
        name: "{{ change.name }}"
      register: cv_show_cc_by_name

    - name: "Print out the change control"
      debug:
        msg: "{{ cv_show_cc_by_name }}" 
```

Sample output:
```
PLAY [CVP Change Control Test approve_and_execute] **********************************************************************************************************

TASK [Create a change control on cv_server] *****************************************************************************************************************
changed: [cv_server]

TASK [Approve and Execute a change control on cv_server using change_id field] ******************************************************************************
changed: [cv_server]

TASK [Show the change control on cv_server using CC name] ***************************************************************************************************
ok: [cv_server]

TASK [Print out the change control] *************************************************************************************************************************
ok: [cv_server] => 
  msg:
    changed: false
    data:
      'change_controls:':
      - time: '2022-09-30T02:30:33.864247041Z'
        value:
          approve:
            notes: Approved via Ansible Playbook
            time: '2022-09-30T02:30:33.517301741Z'
            user: cvpadmin
            value: true
          change:
            name: Leaf 1A Change Control
            notes: Created via playbook
            rootStageId: cc7cc6ea-211c-44a3-9f5d-4f0a82d94210
            stages:
              values:
                13b084a3-31f2-4ea1-8cfd-373161f4c339:
                  name: LEAF1A_MLAG_Health
                  rows:
                    values:
                    - values:
                      - 5ed3b3f9-f04e-4462-831d-42c7cafce40a
                  status: STAGE_STATUS_RUNNING
                5ed3b3f9-f04e-4462-831d-42c7cafce40a:
                  action:
                    args:
                      values:
                        DeviceID: SN-DC1-POD1-LEAF1A
                    name: mlaghealthcheck
                  name: Check_LEAF1A_MLAG_Health
                  rows: {}
                  status: STAGE_STATUS_RUNNING
                cc7cc6ea-211c-44a3-9f5d-4f0a82d94210:
                  name: Leaf 1A Change Control root stage
                  rows:
                    values:
                    - values:
                      - 13b084a3-31f2-4ea1-8cfd-373161f4c339
                  status: STAGE_STATUS_RUNNING
            time: '2022-09-30T02:30:30.995184335Z'
            user: cvpadmin
          key:
            id: 65cb69a2-e964-465c-bfa6-60c78ced9e1b
          start:
            time: '2022-09-30T02:30:33.762347718Z'
            user: cvpadmin
            value: true
          status: CHANGE_CONTROL_STATUS_RUNNING
    failed: false

PLAY RECAP **************************************************************************************************************************************************
cv_server                  : ok=4    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```



## Checklist

### User Checklist


- N/A

### Repository Checklist
